### PR TITLE
sshd hardening: added more and documentation

### PR DESCRIPTION
**Pull request by boredsquirrel**
I hope this format is okay. I found some that I think are missing, and I also added documentation (do not edit this file, you dont get updates).

Also I added many commented out configs people might want to add, but that would break first setup or some use cases.

### CODE REVIEW

File: config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf [hunk 1]
```diff
@@ -1,3 +1,9 @@
+# Secure sshd Template
+
+# If you want to add configurations, instead of editing this file,
+# create another one in the same directory.
```
Comment (RoyalOughtness) [inline config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf]: This is almost good, but it should also say that the new file needs to be alphabetically later.

Also, can you put this in the FAQ instead?

File: config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf [hunk 2]
```diff
@@ -20,4 +26,32 @@ TCPKeepAlive no
 ClientAliveCountMax 2
 
 # Forbid sshing as root
-PermitRootLogin no
\ No newline at end of file
+PermitRootLogin no
+
+# Enable Pubkey Authentification
```
Comment (RoyalOughtness) [inline config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf]: is this not default?

File: config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf [hunk 3]
```diff
@@ -20,4 +26,32 @@ TCPKeepAlive no
...
+PubkeyAuthentication yes
+
+# Block empty passwords
+PermitEmptyPasswords no
+
+# Allow only 256bit AES encryption
```
Comment (RoyalOughtness) [inline config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf]: why is this needed?

File: config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf [hunk 4]
```diff
@@ -20,4 +26,32 @@ TCPKeepAlive no
...
+Ciphers aes256-ctr
+
+# Allow only 512bit SHA-2 HMAC
+MACs hmac-sha2-512
```
Comment (RoyalOughtness) [inline config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf]: why is this needed?

**Review by RoyalOughtness — CHANGES_REQUESTED** [review]

File: config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf (hunk 1 above)
Comment (boredsquirrel) [reply config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf]: yes, I suppose having a central location is good. I came to the same idea but will not add this there, just add this warning (And I think we should add it on every such file?)

then add a FAQ entry on how to add supplementary and overwrite configs.

File: config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf (hunk 3 above)
Comment (boredsquirrel) [reply config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf]: default should be

```
# Ciphers aes128-ctr,aes192-ctr,aes256-ctr
# MACs hmac-sha2-256,hmac-sha2-512
```

so this only accepts the most secure ones. I read somewhere that modern ssh clients default to 256 anyways.

File: config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf (hunk 1 above)
Comment (RoyalOughtness) [reply config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf]: Just the FAQ please to avoid redundancy

File: config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf (hunk 4 above)
Comment (boredsquirrel) [reply config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf]: the Message Authentication Codes are used by ssh for integrity verification of transmitted data, this should protect against some brute force attacks

https://security.stackexchange.com/a/117104

In theory these are not the only secure ones though, so adding the others will be better than random clients not working and people using less secure methods like email.

File: config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf (hunk 2 above)
Comment (boredsquirrel) [reply config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf]: I could remove it to keep aligned with "as upstream as possible"

File: config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf (hunk 4 above)
Comment (boredsquirrel) [reply config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf]: okay I changed those again. Added the cyphers from the post, and allowed the standard ones too for following the last mentioned reason.

Added the most secure config as commented out below, for people that want it.

Comment (boredsquirrel) [reply config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf]: I am actually very unsure if we should block everything but the most secure protocols, with the possibility of breaking clients (add a note that they should use a different client) or keep those protocols.

File: config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf (hunk 3 above)
Comment (RoyalOughtness) [reply config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf]: 256 isn't "more secure" per se

https://security.stackexchange.com/posts/19762/revisions

128 is fine.

Comment (RoyalOughtness) [reply config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf]: Please remove the cipher restrictions and mac restrictions

File: config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf [hunk 5]
```diff
@@ -20,4 +20,33 @@ TCPKeepAlive no
...
+# # Allow only certain users
+# AllowUsers username1 username2
+#
+# # Block TCP Forwarding
+# AllowTcpForwarding no
+#
+# # Disconnect clients without successful login after x seconds
+# LoginGraceTime x
+#
+# # Enforce strongest Ciphers and MACs (may break some clients)
```
Comment (RoyalOughtness) [inline config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf]: This line is misleading as mentioned previously

File: config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf [hunk 6]
```diff
@@ -20,4 +20,23 @@ TCPKeepAlive no
...
+# write them into another file in this directory
+#
+# # Forbid Password Authentication
+# PasswordAuthentication no
+#
+# # Allow only certain users
+# AllowUsers username1 username2
+#
+# # Block TCP Forwarding
+# AllowTcpForwarding no
```
Comment (RoyalOughtness) [inline config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf:39]: Most of these recommendations are dubious.

From the docs:

```
  AllowTcpForwarding
        Specifies whether TCP forwarding is permitted.
        The default is “yes”.  Note that disabling TCP
        forwarding does not improve security unless users
        are also denied shell access, as they can always
        install their own forwarders.
```

File: config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf [hunk 7]
```diff
@@ -20,4 +20,23 @@ TCPKeepAlive no
...
+#
+# # Disconnect clients without successful login after x seconds
+# LoginGraceTime x
```
Comment (RoyalOughtness) [inline config/files/server/usr/etc/ssh/sshd_config.d/99-hardening.conf:42]: This would be better accomplished with something like fail2ban

**Comment by pm4rcin** [comment]
Here's something that I have followed personally. https://infosec.mozilla.org/guidelines/openssh

**Comment by RoyalOughtness** [comment]
> Here's something that I have followed personally. https://infosec.mozilla.org/guidelines/openssh

If there are specific configuration changes you'd like to see, please open a pr!

In the meantime, this PR no longer has much substance so I'll be closing it.
